tests/unit/conftest.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Fixtures for Jenkins-k8s-operator charm unit tests."""
+
+from io import StringIO
+from pathlib import Path
+from secrets import token_hex
+from typing import Any, BinaryIO, Optional, TextIO, Union
+from unittest.mock import MagicMock
+
+import pytest
+import requests
+from ops.model import Container
+from ops.testing import Harness
+
+from charm import JenkinsK8SOperatorCharm
+from jenkins import JENKINS_PASSWORD_FILE_PATH
+from types_ import Credentials
+
+from .helpers import make_relative_to_path
+from .types_ import ContainerWithPath
+
+
+@pytest.fixture(scope="function", name="harness")
+def harness_fixture():
+    """Enable ops test framework harness."""
+    harness = Harness(JenkinsK8SOperatorCharm)
+    yield harness
+    harness.cleanup()
+
+
+@pytest.fixture(scope="function", name="jenkins_version")
+def jenkins_version_fixture():
+    """Jenkins version fixture."""
+    return "2.400"
+
+
+@pytest.fixture(scope="function", name="mocked_get_request")
+def mocked_get_request_fixture(jenkins_version: str):
+    """Mock get request with given status code."""
+
+    def mocked_get(url: str, status_code: int = 200, **kwargs: Any):
+        """Mock get request with predefined status code.
+
+        Args:
+            url: URL placeholder parameter to match original function signature.
+            status_code: Status code of the returned response.
+            kwargs: Other keyword arguments to match function signature.
+
+        Returns:
+            Mocked response.
+        """
+        del url, kwargs
+        response = requests.Response()
+        response.status_code = status_code
+        response.headers["X-Jenkins"] = jenkins_version
+        return response
+
+    return mocked_get
+
+
+@pytest.fixture(scope="function", name="admin_credentials")
+def admin_credentials_fixture() -> Credentials:
+    """Admin credentials for Jenkins."""
+    return Credentials(username="admin", password=token_hex(16))
+
+
+@pytest.fixture(scope="function", name="container_tmppath")
+def container_tmppath_fixture(tmp_path: Path, admin_credentials: Credentials) -> Path:
+    """Temporary directory structure for Jenkins container."""
+    # if the path is an absolute path starting at root / directory, we must make it relative
+    # path, otherwise the overloaded path append operator (/) doesn't work.
+    initial_password_path = make_relative_to_path(tmp_path, JENKINS_PASSWORD_FILE_PATH)
+    initial_password_path.parent.mkdir(exist_ok=True, parents=True)
+    initial_password_path.write_text(admin_credentials.password, encoding="utf-8", newline="\n")
+    return tmp_path
+
+
+@pytest.fixture(scope="function", name="mocked_container")
+def mocked_container_fixture(
```

Comment:
Is this the same as what we had in the 12-factor charm? Perhaps we can put it as a pattern somewhere in the contributing guide repo?

Reply:
And mentioned where it came from here so we don't review the same code multiple times 🙂

Reply:
Refactored to match 12-factor charm testing method, will add adding to the contributing guide for TODO for tomorrow!